File: kinova_gen3/gen3_interface.py
import threading

from typing import Any, ClassVar, List, Optional
from types import SimpleNamespace
from pydrake.all import *
from kortex_api.TCPTransport import TCPTransport
from kortex_api.RouterClient import RouterClient, RouterClientSendOptions
from kortex_api.SessionManager import SessionManager

from kortex_api.autogen.client_stubs.DeviceConfigClientRpc import DeviceConfigClient
from kortex_api.autogen.client_stubs.DeviceManagerClientRpc import DeviceManagerClient
from kortex_api.autogen.client_stubs.VisionConfigClientRpc import VisionConfigClient

from kortex_api.autogen.client_stubs.BaseClientRpc import BaseClient
from kortex_api.autogen.client_stubs.BaseCyclicClientRpc import BaseCyclicClient

from kortex_api.autogen.messages import (
    DeviceConfig_pb2,
    Session_pb2,
    Base_pb2,
    VisionConfig_pb2,
)
import sys
from .gen3_constants import (
    Gen3ControlMode,
    kGen3ArmNumJoints,
    Gen3NamedPosition,
    kGen3NamedPositions,
)

from common import *


class Gen3InterfaceConfig:
    __fields__: ClassVar[tuple] = (
        SimpleNamespace(name="port", type=str),
        SimpleNamespace(name="ip_address", type=str),
    )
    port: str
    ip_address: str


class Gen3HardwareInterface(LeafSystem):
    def __init__(self, ip_address, port, hand_model_name, sim_plant: MultibodyPlant):
        LeafSystem.__init__(self)
        self.hand_model_name = hand_model_name
        self.sim_plant = sim_plant
        # World to robot base
        self.X_WB: RigidTransform = self.sim_plant.CalcRelativeTransform(
            self.sim_plant.CreateDefaultContext(),
            self.sim_plant.world_frame(),
            self.sim_plant.GetFrameByName(
                "base_link", self.sim_plant.GetModelInstanceByName("gen3")
            ),
        )
        self.X_WB_inv: RigidTransform = self.X_WB.inverse()

        self.root_ctx = None

        # 7 values are one of the following:
        #   joint positions
        #   joint velocities
        #   pose (rpy, xyz)
        #   twist (rpy, xyz, 0)
        self.arm_command_input_port = self.DeclareVectorInputPort("command", 7)
        self.control_mode_input_port = self.DeclareAbstractInputPort(
            "control_mode", AbstractValue.Make(Gen3ControlMode.kPose)
        )

        if hand_model_name == "2f_85":
            self.gripper_command_input_port = self.DeclareVectorInputPort(
                "2f_85.command", 1
            )
        # self.DeclareVectorOutputPort(
        #     "position_measured", BasicVector(kGen3ArmNumJoints), self.CalcArmPosition
        # )
        # self.DeclareVectorOutputPort(
        #     "velocity_measured", BasicVector(kGen3ArmNumJoints), self.CalcArmVelocity
        # )
        # self.DeclareVectorOutputPort(
        #     "torque_measured", BasicVector(kGen3ArmNumJoints), self.CalcArmTorque
        # )
        # self.DeclareVectorOutputPort(
        #     "state_estimated", BasicVector(kGen3ArmNumJoints * 2), self.CalcState
        # )
        self.DeclareVectorOutputPort(
            "pose_measured",
            BasicVector(6),
            self.CalcEndEffectorPose,
            {self.time_ticket()},
        )
        self.DeclarePeriodicUnrestrictedUpdateEvent(1 / 40, 0, self.Integrate)

        self.last_feedback_time = -np.inf
        self.feedback = None

        logging.info("Connecting to gen3 arm...")
        self.transport = TCPTransport()
        self.router = RouterClient(self.transport, RouterClient.basicErrorCallback)
        self.transport.connect(ip_address, int(port))
        # Create session
        session_info = Session_pb2.CreateSessionInfo()
        session_info.username = "admin"
        session_info.password = "admin"
        session_info.session_inactivity_timeout = 6000  # (milliseconds)
        session_info.connection_inactivity_timeout = 2000  # (milliseconds)

        self.session_manager = SessionManager(self.router)
        self.session_manager.CreateSession(session_info)

        # Create required services
        device_config = DeviceConfigClient(self.router)
        self.base = BaseClient(self.router)
        self.base_cyclic = BaseCyclicClient(self.router)

        if self.base.GetArmState().active_state != Base_pb2.ARMSTATE_SERVOING_READY:
            logging.error(self.base.GetArmState())
            raise RuntimeError(
                "Arm not in ready state. Clear any faults before trying again."
            )

    def CleanUp(self):
        logging.info("\nClosing Hardware Connection...")

        if self.session_manager is not None:
            router_options = RouterClientSendOptions()
            router_options.timeout_ms = 1000

            self.session_manager.CloseSession()

        self.transport.disconnect()
        logging.info("Hardware Connection Closed.")

    def GetFeedback(self, current_time):
        """
        Sets self.feedback with the latest data from the controller.
        We also indicate the time at which this feedback was set, in an
        effor to reduce redudnant calls to self.base_cyclic.RefreshFeedback(),
        which take about 25ms each.
        """
        self.feedback = self.base_cyclic.RefreshFeedback()
        self.last_feedback_time = current_time

    def check_for_end_or_abort(self, e):
        """
        Return a closure checking for END or ABORT notifications

        Arguments:
        e -- event to signal when the action is completed
            (will be set when an END or ABORT occurs)
        """

        def check(notification, e=e):
            if (
                notification.action_event == Base_pb2.ACTION_END
                or notification.action_event == Base_pb2.ACTION_ABORT
            ):
                e.set()

        return check

    def go_to_named_position(self, named_position: Gen3NamedPosition):
        self.SendPositionCommand(kGen3NamedPositions[named_position])

    def SendPoseCommand(self, position: np.ndarray, rpy: RollPitchYaw):
        action = Base_pb2.Action()
        action.name = "End-effector pose command"
        action.application_data = ""

        cartesian_pose = action.reach_pose.target_pose
        cartesian_pose.theta_x = np.degrees(rpy.roll_angle())
        cartesian_pose.theta_y = np.degrees(rpy.pitch_angle())
        cartesian_pose.theta_z = np.degrees(rpy.yaw_angle())
        cartesian_pose.x = position[0]
        cartesian_pose.y = position[1]
        cartesian_pose.z = position[2]

        e = threading.Event()
        notification_handle = self.base.OnNotificationActionTopic(
            self.check_for_end_or_abort(e), Base_pb2.NotificationOptions()
        )

        self.base.ExecuteAction(action)

        TIMEOUT_DURATION = 20  # seconds
        finished = e.wait(TIMEOUT_DURATION)
        self.base.Unsubscribe(notification_handle)

    def SendTwistCommand(self, cmd_twist: np.ndarray):
        command = Base_pb2.TwistCommand()
        command.reference_frame = Base_pb2.CARTESIAN_REFERENCE_FRAME_BASE
        command.duration = 0
        twist = command.twist
        twist.angular_x = np.degrees(cmd_twist[0])
        twist.angular_y = np.degrees(cmd_twist[1])
        twist.angular_z = np.degrees(cmd_twist[2])
        twist.linear_x = cmd_twist[3]
        twist.linear_y = cmd_twist[4]
        twist.linear_z = cmd_twist[5]

        # Note: this API call takes about 25ms
        self.base.SendTwistCommand(command)

    def SendWrenchCommand(self, cmd_wrench):
        command = Base_pb2.WrenchCommand()
        command.reference_frame = Base_pb2.CARTESIAN_REFERENCE_FRAME_BASE
        command.duration = 0

        wrench = command.wrench
        wrench.torque_x = cmd_wrench[0]
        wrench.torque_y = cmd_wrench[1]
        wrench.torque_z = cmd_wrench[2]
        wrench.force_x = cmd_wrench[3]
        wrench.force_y = cmd_wrench[4]
        wrench.force_z = cmd_wrench[5]

        self.base.SendWrenchCommand(command)

    def SendPositionCommand(self, joint_positions: np.ndarray):
        action = Base_pb2.Action()
        action.name = ""
        action.application_data = ""

        actuator_count = self.base.GetActuatorCount()

        # Place arm straight up
        for joint_id in range(actuator_count.count):
            joint_angle = action.reach_joint_angles.joint_angles.joint_angles.add()
            joint_angle.joint_identifier = joint_id
            joint_angle.value = np.degrees(joint_positions[joint_id])

        e = threading.Event()
        notification_handle = self.base.OnNotificationActionTopic(
            self.check_for_end_or_abort(e), Base_pb2.NotificationOptions()
        )

        self.base.ExecuteAction(action)
        TIMEOUT_DURATION = 20  # seconds
        finished = e.wait(TIMEOUT_DURATION)
        self.base.Unsubscribe(notification_handle)

    def SendVelocityCommand(self, joint_velocities: np.ndarray):
        joint_speeds = Base_pb2.JointSpeeds()

        actuator_count = self.base.GetActuatorCount().count
        assert actuator_count == len(joint_velocities)
        for i, speed in enumerate(joint_velocities):
            joint_speed = joint_speeds.joint_speeds.add()
            joint_speed.joint_identifier = i
            joint_speed.value = speed
            joint_speed.duration = 0

        self.base.SendJointSpeedsCommand(joint_speeds)

    def Integrate(self, context: Context, discrete_state: DiscreteValues):
        # def DoCalcTimeDerivatives(self, context, continuous_state):
        # TODO this is a bad way to do this:
        ## ------
        t = context.get_time()
        if self.last_feedback_time != t:
            self.GetFeedback(t)

        q = np.zeros(kGen3ArmNumJoints)
        for i in range(kGen3ArmNumJoints):
            q[i] = np.radians(self.feedback.actuators[i].position)

        self.sim_plant.SetPositions(
            self.sim_plant.GetMyMutableContextFromRoot(self.root_ctx),
            self.sim_plant.GetModelInstanceByName("gen3"),
            q,
        )
        self.sim_plant.SetPositions(
            self.sim_plant.GetMyMutableContextFromRoot(self.root_ctx),
            self.sim_plant.GetModelInstanceByName("2f_85"),
            np.zeros(6),
        )
        ## ------
        if self.hand_model_name == "2f_85":
            gripper_command = Base_pb2.GripperCommand()
            gripper_command.mode = Base_pb2.GRIPPER_SPEED

            finger = gripper_command.gripper.finger.add()
            finger.finger_identifier = 1
            finger.value = self.gripper_command_input_port.Eval(context)[0]

            self.base.SendGripperCommand(gripper_command)

        control_mode: Gen3ControlMode = self.control_mode_input_port.Eval(context)

        command: BasicVector = self.arm_command_input_port.Eval(context)

        match control_mode:
            case Gen3ControlMode.kPosition:
                self.SendPositionCommand(command)
            case Gen3ControlMode.kVelocity:
                self.SendVelocityCommand(command)
            case Gen3ControlMode.kPose:
                translation = command[3:-1]
                rpy = RollPitchYaw(command[:3])
                # X_WE = X_WB @ X_BE
                X_BE = self.X_WB_inv = RigidTransform(rpy.ToQuaternion(), translation)
                self.SendPoseCommand(
                    X_BE.translation(), X_BE.rotation().ToRollPitchYaw()
                )
            case Gen3ControlMode.kTwist:
                self.SendTwistCommand(command[:6])
            case _:
                raise RuntimeError(f"Unsupported control mode {control_mode}")

    def CalcArmPosition(self, context, output):
        """
        Compute the current joint angles and send as output.
        """
        # Get feedback from the base, but only if we haven't already this timestep
        t = context.get_time()
        if self.last_feedback_time != t:
            self.GetFeedback(t)

        q = np.zeros(kGen3ArmNumJoints)
        for i in range(kGen3ArmNumJoints):
            q[i] = np.radians(
                self.feedback.actuators[i].position
            )  # Kortex provides joint angles
            # in degrees for some reason
        output.SetFromVector(q)

    def CalcArmVelocity(self, context, output):
        """
        Compute the current joint velocities and send as output.
        """
        t = context.get_time()
        if self.last_feedback_time != t:
            self.GetFeedback(t)

        qd = np.zeros(kGen3ArmNumJoints)
        for i in range(kGen3ArmNumJoints):
            qd[i] = np.radians(
                self.feedback.actuators[i].velocity
            )  # Kortex provides joint angles
            # in degrees for some reason
        output.SetFromVector(qd)

    def CalcArmTorque(self, context, output):
        t = context.get_time()
        if self.last_feedback_time != t:
            self.GetFeedback(t)

        tau = np.zeros(kGen3ArmNumJoints)
        for i in range(kGen3ArmNumJoints):
            tau[i] = np.radians(self.feedback.actuators[i].torque)  # in Nm

        output.SetFromVector(tau)

    def CalcState(self, context, output):
        t = context.get_time()
        if self.last_feedback_time != t:
            self.GetFeedback(t)

        q = np.zeros(kGen3ArmNumJoints)
        for i in range(kGen3ArmNumJoints):
            q[i] = np.radians(
                self.feedback.actuators[i].position
            )  # Kortex provides joint angles
        qd = np.zeros(kGen3ArmNumJoints)
        for i in range(kGen3ArmNumJoints):
            qd[i] = np.radians(
                self.feedback.actuators[i].velocity
            )  # Kortex provides joint angles
            # in degrees for some reason

        output.SetFromVector(np.concatenate((q, qd), axis=0))

    def CalcEndEffectorPose(self, context, output):
        """
        Compute the current end-effector pose and send as output.
        """
        t = context.get_time()
        if self.last_feedback_time != t:
            self.GetFeedback(t)

        ee_pose = np.zeros(6)
        ee_pose[0] = np.radians(self.feedback.base.tool_pose_theta_x)
        ee_pose[1] = np.radians(self.feedback.base.tool_pose_theta_y)
        ee_pose[2] = np.radians(self.feedback.base.tool_pose_theta_z)
        ee_pose[3] = self.feedback.base.tool_pose_x
        ee_pose[4] = self.feedback.base.tool_pose_y
        ee_pose[5] = self.feedback.base.tool_pose_z

        # pose : RigidTransform = self.X_WB @ RigidTransform(RollPitchYaw(ee_pose[:3]).ToQuaternion(), ee_pose[3:])
        # o = np.zeros(7)
        # # o[:4] = pose.rotation().ToQuaternion().wxyz()
        # o[:3] = np.array([0, 3.14, 0])
        # o[3:-1] = pose.translation()
        output.SetFromVector(ee_pose)

File: kinova_gen3/gen3_interface.cc
#include "kinova_gen3/gen3_interface.h"

namespace piplup
{
    namespace kinova_gen3
    {
        using namespace drake;
        using systems::BasicVector;
        namespace k_api = Kinova::Api;
        Gen3HardwareInterface::Gen3HardwareInterface(std::string ip_address,
                                                     std::string port,
                                                     Gen3HandType hand_type)
          : hand_type_(hand_type)
        {
            // State
            pose_measured_state_index_ = DeclareDiscreteState(6);
            twist_measured_state_index_ = DeclareDiscreteState(6);
            wrench_measured_state_index_ = DeclareDiscreteState(6);
            position_measured_state_index_ = DeclareDiscreteState(7);
            velocity_measured_state_index_ = DeclareDiscreteState(7);
            torque_measured_state_index_ = DeclareDiscreteState(7);
            feedback_future_index_ = DeclareAbstractState(
                Value<std::shared_future<k_api::BaseCyclic::Feedback>>());
            gripper_future_index_ =
                DeclareAbstractState(Value<std::shared_future<void>>());
            arm_future_index_ = DeclareAbstractState(Value<std::shared_future<void>>());

            // DeclareAbstractState(Value<std::future<k_api::BaseCyclic::Feedback>>());
            // Input Ports
            command_port_ = &DeclareVectorInputPort("command", 7);
            control_mode_port_ =
                &DeclareAbstractInputPort("control_mode", Value<Gen3ControlMode>());

            if(hand_type_ == Gen3HandType::k2f85)
            {
                hand_command_port_ =
                    &DeclareVectorInputPort("2f_85.command", BasicVector<double>({0.0}));
            }

            // Output Ports
            DeclareStateOutputPort("pose_measured", pose_measured_state_index_);
            DeclareStateOutputPort("twist_measured", twist_measured_state_index_);
            DeclareStateOutputPort("wrench_measured", wrench_measured_state_index_);
            DeclareStateOutputPort("position_measured", position_measured_state_index_);
            DeclareStateOutputPort("velocity_measured", velocity_measured_state_index_);
            DeclareStateOutputPort("torque_measured", torque_measured_state_index_);

            // Update Events
            double hz = 20.0;
            DeclareInitializationUnrestrictedUpdateEvent(
                &Gen3HardwareInterface::Initialize);
            // DeclarePeriodicPublishEvent(1 / hz, 0.0,
            // &Gen3HardwareInterface::CalcUpdate);
            DeclarePeriodicUnrestrictedUpdateEvent(
                1 / hz, 0.0, &Gen3HardwareInterface::CalcUpdate);
            auto error_callback = [](k_api::KError err) {
                cout << "_________ callback error _________" << err.toString();
            };
            drake::log()->info("Connecting to transport client...");
            transport_ = new k_api::TransportClientTcp();
            router_ = new k_api::RouterClient(transport_, error_callback);
            transport_->connect("192.168.1.10", 10000);

            // Set session data connection information
            auto create_session_info = k_api::Session::CreateSessionInfo();
            create_session_info.set_username("admin");
            create_session_info.set_password("admin");
            create_session_info.set_session_inactivity_timeout(60000);   // (milliseconds)
            create_session_info.set_connection_inactivity_timeout(2000); // (milliseconds)

            // Session manager service wrapper
            drake::log()->info("Creating session for communication");
            session_manager_ = new k_api::SessionManager(router_);
            session_manager_->CreateSession(create_session_info);
            drake::log()->info("Session created");
            // Create services
            base_ = new k_api::Base::BaseClient(router_);
            base_cyclic_ = new k_api::BaseCyclic::BaseCyclicClient(router_);
        }
        systems::EventStatus Gen3HardwareInterface::Initialize(
            const systems::Context<double> & context,
            systems::State<double> * state) const
        {
            auto servoing_mode = k_api::Base::ServoingModeInformation();
            servoing_mode.set_servoing_mode(
                k_api::Base::ServoingMode::SINGLE_LEVEL_SERVOING);
            base_->SetServoingMode(servoing_mode);
            if(base_->GetArmState().active_state()
               != k_api::Common::ArmState::ARMSTATE_SERVOING_READY)
            {
                return systems::EventStatus::Failed(this, "Arm not in ready state.");
            }
            return systems::EventStatus::Succeeded();
        }

        void Gen3HardwareInterface::DoCalcNextUpdateTime(
            const systems::Context<double> & context,
            systems::CompositeEventCollection<double> * events, double * time) const
        {
            systems::LeafSystem<double>::DoCalcNextUpdateTime(context, events, time);
            auto callback = [](const System<double> & system,
                               const systems::Context<double> & callback_context,
                               const systems::UnrestrictedUpdateEvent<double> &,
                               systems::State<double> * callback_state) {
                const auto & self = dynamic_cast<const Gen3HardwareInterface &>(system);
                auto & feedback_future = callback_state->get_abstract_state<
                    std::shared_future<k_api::BaseCyclic::Feedback>>(
                    self.feedback_future_index_);
                if(!feedback_future.valid()
                   || feedback_future.wait_for(std::chrono::seconds(0))
                          == std::future_status::ready)
                {
                    drake::log()->debug("Feedback Time: {}", callback_context.get_time());
                    if (feedback_future.valid()) {
                        self.UpdateState(feedback_future.get(), callback_state);
                    }
                    auto future = self.base_cyclic_->RefreshFeedback_async().share();
                    callback_state->get_mutable_abstract_state<
                        std::shared_future<k_api::BaseCyclic::Feedback>>(
                        self.feedback_future_index_) = future;
                }
                else
                {
                    return systems::EventStatus::DidNothing();
                }
                return systems::EventStatus::Succeeded();
            };
            *time = context.get_time() + 1.0 / 20.0;
            systems::EventCollection<systems::UnrestrictedUpdateEvent<double>> &
                uu_events = events->get_mutable_unrestricted_update_events();
            uu_events.AddEvent(systems::UnrestrictedUpdateEvent<double>(
                systems::TriggerType::kTimed, callback));
        }

        void Gen3HardwareInterface::CalcUpdate(const systems::Context<double> & context,
                                               systems::State<double> * state) const
        {
            auto gripper_future = state->get_abstract_state<std::shared_future<void>>(
                gripper_future_index_);
            auto arm_future =
                state->get_abstract_state<std::shared_future<void>>(arm_future_index_);
            if(hand_type_ == Gen3HandType::k2f85
               && (!gripper_future.valid()
                   || gripper_future.wait_for(std::chrono::seconds(0))
                          == std::future_status::ready))
            {
                const auto & hand_command =
                    hand_command_port_->Eval<BasicVector<double>>(context);

                k_api::Base::GripperCommand gripper_command;
                auto gripper_command_mode = k_api::Base::GRIPPER_POSITION;
                gripper_command.set_mode(gripper_command_mode);
                drake::log()->debug("Sending Gripper Command {} \t Mode: {}",
                                    hand_command,
                                    gripper_command_mode);
                auto finger = gripper_command.mutable_gripper()->add_finger();
                finger->set_finger_identifier(1);
                finger->set_value(hand_command[0]);

                auto future = base_->SendGripperCommand_async(gripper_command).share();
                state->get_mutable_abstract_state<std::shared_future<void>>(
                    gripper_future_index_) = future;
            }

            const auto & command = command_port_->Eval<BasicVector<double>>(context);
            const auto & control_mode =
                control_mode_port_->Eval<Gen3ControlMode>(context);

            if(!arm_future.valid()
               || arm_future.wait_for(std::chrono::seconds(0))
                      == std::future_status::ready)
            {

                switch(control_mode)
                {
                case Gen3ControlMode::kTwist:
                {
                    drake::log()->debug("Sending Twist Command {}", command);
                    auto twist_command = k_api::Base::TwistCommand();
                    twist_command.set_reference_frame(
                        k_api::Common::CARTESIAN_REFERENCE_FRAME_BASE);
                    twist_command.set_duration(0);
                    auto twist = twist_command.mutable_twist();
                    twist->set_angular_x(command[0]);
                    twist->set_angular_y(command[1]);
                    twist->set_angular_z(command[2]);
                    twist->set_linear_x(command[3]);
                    twist->set_linear_y(command[4]);
                    twist->set_linear_z(command[5]);
                    auto future = base_->SendTwistCommand_async(twist_command).share();
                    state->get_mutable_abstract_state<std::shared_future<void>>(
                        arm_future_index_) = future;
                    break;
                }
                default:
                    drake::log()->error("Gen3 control mode not implemented");
                }
            }
        }

        void Gen3HardwareInterface::UpdateState(
            const k_api::BaseCyclic::Feedback & feedback,
            systems::State<double> * state) const
        {
            for(int i = 0; i < feedback.actuators_size(); i++)
            {
                // drake::log()->debug(
                //     "Actuator {} \t Position: {} \t Velocity: {} \t Torque: {}",
                //     i,
                //     feedback.actuators(i).position(),
                //     feedback.actuators(i).velocity(),
                //     feedback.actuators(i).torque());
                state->get_mutable_discrete_state(position_measured_state_index_)
                    .SetAtIndex(i, feedback.actuators(i).position() * (M_PI / 180.0));
                state->get_mutable_discrete_state(velocity_measured_state_index_)
                    .SetAtIndex(i, feedback.actuators(i).velocity() * (M_PI / 180.0));
                state->get_mutable_discrete_state(torque_measured_state_index_)
                    .SetAtIndex(i, feedback.actuators(i).torque() * (M_PI / 180.0));
            }
            state->get_mutable_discrete_state(pose_measured_state_index_)
                .SetAtIndex(0, feedback.base().tool_pose_theta_x() * (M_PI / 180.0));
            state->get_mutable_discrete_state(pose_measured_state_index_)
                .SetAtIndex(1, feedback.base().tool_pose_theta_y() * (M_PI / 180.0));
            state->get_mutable_discrete_state(pose_measured_state_index_)
                .SetAtIndex(2, feedback.base().tool_pose_theta_z() * (M_PI / 180.0));
            state->get_mutable_discrete_state(pose_measured_state_index_)
                .SetAtIndex(3, feedback.base().tool_pose_x());
            state->get_mutable_discrete_state(pose_measured_state_index_)
                .SetAtIndex(4, feedback.base().tool_pose_y());
            state->get_mutable_discrete_state(pose_measured_state_index_)
                .SetAtIndex(5, feedback.base().tool_pose_z());

            state->get_mutable_discrete_state(twist_measured_state_index_)
                .SetAtIndex(0, feedback.base().tool_twist_angular_x() * (M_PI / 180.0));
            state->get_mutable_discrete_state(twist_measured_state_index_)
                .SetAtIndex(1, feedback.base().tool_twist_angular_y() * (M_PI / 180.0));
            state->get_mutable_discrete_state(twist_measured_state_index_)
                .SetAtIndex(2, feedback.base().tool_twist_angular_z() * (M_PI / 180.0));
            state->get_mutable_discrete_state(twist_measured_state_index_)
                .SetAtIndex(3, feedback.base().tool_twist_linear_x());
            state->get_mutable_discrete_state(twist_measured_state_index_)
                .SetAtIndex(4, feedback.base().tool_twist_linear_y());
            state->get_mutable_discrete_state(twist_measured_state_index_)
                .SetAtIndex(5, feedback.base().tool_twist_linear_z());

            state->get_mutable_discrete_state(wrench_measured_state_index_)
                .SetAtIndex(
                    0, feedback.base().tool_external_wrench_torque_x() * (M_PI / 180.0));
            state->get_mutable_discrete_state(wrench_measured_state_index_)
                .SetAtIndex(
                    1, feedback.base().tool_external_wrench_torque_y() * (M_PI / 180.0));
            state->get_mutable_discrete_state(wrench_measured_state_index_)
                .SetAtIndex(
                    2, feedback.base().tool_external_wrench_torque_z() * (M_PI / 180.0));
            state->get_mutable_discrete_state(wrench_measured_state_index_)
                .SetAtIndex(3, feedback.base().tool_external_wrench_force_x());
            state->get_mutable_discrete_state(wrench_measured_state_index_)
                .SetAtIndex(4, feedback.base().tool_external_wrench_force_y());
            state->get_mutable_discrete_state(wrench_measured_state_index_)
                .SetAtIndex(5, feedback.base().tool_external_wrench_force_z());
        }

        Gen3HardwareInterface::~Gen3HardwareInterface()
        {
            drake::log()->info("Closing Kinova Interface...");
            session_manager_->CloseSession();

            // Deactivate the router and cleanly disconnect from the transport object
            router_->SetActivationStatus(false);
            transport_->disconnect();

            // Destroy the API
            delete base_;
            delete session_manager_;
            delete router_;
            delete transport_;
            drake::log()->info("Kinova Interface Closed");
        }
    } // namespace kinova_gen3
} // namespace piplup